File: modules/echo-audio/android/src/main/java/expo/modules/echoaudio/EchoAudioModule.kt
package expo.modules.echoaudio

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel

/**
 * EchoAudio — Android voice pipeline.
 *
 * Layout: AudioCapture -> VAD -> OpusEncoder -> emit 'onCapturedFrame'
 *         pushIncomingFrame -> OpusDecoder -> JitterBuffer -> AudioPlayback
 *
 * STAGE 2 status:
 *  - Capture (AudioRecord) and playback (AudioTrack) are real.
 *  - RMS-based VAD is real.
 *  - OpusEncoder / OpusDecoder are stubbed — they pass PCM through. Wire a
 *    JNI Opus binding (e.g. concentus) to harden voice quality and reduce
 *    bandwidth before shipping.
 *  - JitterBuffer is a simple FIFO with a max depth. TODO(stage-2): jitter
 *    estimation + adaptive delay.
 */
class EchoAudioModule : Module() {
  private val scope = CoroutineScope(Dispatchers.Default + SupervisorJob())
  private lateinit var capture: AudioCapture
  private lateinit var playback: AudioPlayback
  private val jitterBuffer = JitterBuffer(maxDepth = 10)
  private val stats = AudioStats()

  override fun definition(): ModuleDefinition = ModuleDefinition {
    Name("EchoAudio")
    Events("onCapturedFrame", "onTalker")

    OnCreate {
      capture = AudioCapture(scope, ::onFrameCaptured, ::onTalker)
      playback = AudioPlayback(scope, jitterBuffer, stats)
    }

    OnDestroy {
      capture.stop()
      playback.stop()
      scope.cancel()
    }

    AsyncFunction("startCapture") { opts: Map<String, Any?> ->
      val mode = (opts["mode"] as? String) ?: "ptt"
      val vadThreshold = (opts["vadThreshold"] as? Number)?.toFloat() ?: 0.05f
      val sampleRate = (opts["sampleRateHz"] as? Number)?.toInt() ?: 16000
      capture.start(mode = mode, vadThreshold = vadThreshold, sampleRate = sampleRate)
      playback.start(sampleRate)
    }

    AsyncFunction("stopCapture") {
      capture.stop()
      playback.stop()
    }

    AsyncFunction("pushIncomingFrame") { frame: Map<String, Any?> ->
      val data = (frame["data"] as? String) ?: return@AsyncFunction
      val ts = (frame["ts"] as? Number)?.toLong() ?: System.currentTimeMillis()
      val duration = (frame["durationMs"] as? Number)?.toInt() ?: 20
      jitterBuffer.push(JitterBuffer.Frame(data, ts, duration))
    }

    AsyncFunction("getStats") {
      mapOf(
        "capturedFrames" to stats.captured,
        "playedFrames" to stats.played,
        "droppedFrames" to stats.dropped,
        "jitterBufferDepth" to jitterBuffer.depth(),
      )
    }
  }

  private fun onFrameCaptured(b64: String, ts: Long, durationMs: Int, voiced: Boolean) {
    stats.captured++
    sendEvent("onCapturedFrame", mapOf(
      "data" to b64,
      "ts" to ts,
      "durationMs" to durationMs,
      "voiced" to voiced,
    ))
  }

  private fun onTalker(active: Boolean) {
    sendEvent("onTalker", mapOf("active" to active))
  }
}

internal class AudioStats {
  @Volatile var captured: Long = 0
  @Volatile var played: Long = 0
  @Volatile var dropped: Long = 0
}

File: modules/echo-audio/ios/EchoAudioModule.swift
import ExpoModulesCore
import AVFoundation

/**
 * EchoAudio — iOS voice pipeline.
 *
 * AVAudioEngine input tap -> 20ms PCM frames -> RMS VAD -> Opus encode ->
 * onCapturedFrame. pushIncomingFrame -> Opus decode -> JitterBuffer ->
 * AVAudioPlayerNode.
 *
 * STAGE 2 status:
 *  - Capture (AVAudioEngine) and playback (AVAudioPlayerNode) are real.
 *  - RMS-based VAD is real.
 *  - Opus encode/decode are stubbed (PCM passthrough).
 *  - JitterBuffer is FIFO.
 */
public class EchoAudioModule: Module {
  private var capture: AudioCapture?
  private var playback: AudioPlayback?
  private let jitter = JitterBuffer(maxDepth: 10)
  private let stats = AudioStats()

  public func definition() -> ModuleDefinition {
    Name("EchoAudio")
    Events("onCapturedFrame", "onTalker")

    OnCreate {
      self.capture = AudioCapture(
        onFrame: { [weak self] b64, ts, durationMs, voiced in
          guard let self else { return }
          self.stats.captured += 1
          self.sendEvent("onCapturedFrame", [
            "data": b64,
            "ts": ts,
            "durationMs": durationMs,
            "voiced": voiced,
          ])
        },
        onTalker: { [weak self] active in
          self?.sendEvent("onTalker", ["active": active])
        }
      )
      self.playback = AudioPlayback(jitter: self.jitter, stats: self.stats)
    }

    OnDestroy {
      self.capture?.stop()
      self.playback?.stop()
    }

    AsyncFunction("startCapture") { (opts: [String: Any]) -> Void in
      let mode = (opts["mode"] as? String) ?? "ptt"
      let vad = (opts["vadThreshold"] as? Double) ?? 0.05
      let sr = (opts["sampleRateHz"] as? Int) ?? 16000
      self.capture?.start(mode: mode, vadThreshold: Float(vad), sampleRate: Double(sr))
      self.playback?.start(sampleRate: Double(sr))
    }

    AsyncFunction("stopCapture") {
      self.capture?.stop()
      self.playback?.stop()
    }

    AsyncFunction("pushIncomingFrame") { (frame: [String: Any]) in
      guard let data = frame["data"] as? String else { return }
      let ts = (frame["ts"] as? Double) ?? Date().timeIntervalSince1970 * 1000
      let dur = (frame["durationMs"] as? Int) ?? 20
      self.jitter.push(JitterFrame(data: data, ts: ts, durationMs: dur))
    }

    AsyncFunction("getStats") { () -> [String: Any] in
      return [
        "capturedFrames": self.stats.captured,
        "playedFrames": self.stats.played,
        "droppedFrames": self.stats.dropped,
        "jitterBufferDepth": self.jitter.depth(),
      ]
    }
  }
}

final class AudioStats {
  var captured: Int = 0
  var played: Int = 0
  var dropped: Int = 0
}
